### ISSUE TITLE
Charm shows as active/idle even when raft configuration is incorrect

### ISSUE THREAD

**Issue by DanielArndt** [canonical/vault-k8s-operator#100]
#### Describe the bug
As a follow-up to #88 -- when this happens there isn't an immediately obvious way to detect that something is wrong.

The charm will remain in active/idle even when raft has reported that it is unhealthy. There should probably be some sort of warning in this case.

Furthermore, it's not obvious what triggers raft to become unhealthy. Even without the bugfix in #88, when I tested manually after the new pod rejoined raft, everything appeared to be healthy. Raft only reached an unhealthy state once I removed the leader unit.

#### To Reproduce

Deploy _an old version of the charm_, before the bugfix in 88.

```
juju add-model raft-cluster-bug
juju deploy vault-k8s --channel edge -n 5 --trust --revision 53
```

Wait for everything to be up and running. Then, configure vault to connect to your vault cluster directly.

Install vault and yq if not already installed

```bash
sudo apt install vault yq
```

```bash
# Set vault token
vault_initialization_secret=$(juju secrets --format=yaml | yq '. | with_entries(select(.*.label == "vault-initialization")) | keys | .[]')
export VAULT_TOKEN=$(juju show-secret ${vault_initialization_secret} --reveal | yq '.[].content.roottoken')

# Extract the vault CA cert and set the path
vault_ca_certificate_secret=$(juju secrets --format=yaml | yq '. | with_entries(select(.*.label == "vault-ca-certificate")) | keys | .[]')
juju show-secret ${vault_ca_certificate_secret} --reveal | yq '.[].content.certificate' > vault_ca.pem
export VAULT_CAPATH=$PWD/vault_ca.pem

# Set the address of vault
export VAULT_ADDR="https://$(juju status vault-k8s --format=yaml | yq '.applications.vault-k8s.address'):8200"
```

Check the raft health

```
vault operator raft autopilot state
```

```console
$ vault operator raft autopilot state
Healthy:                      true
Failure Tolerance:            2
Leader:                       raft-cluster-bug-vault-k8s/0
Voters:
   raft-cluster-bug-vault-k8s/0
   raft-cluster-bug-vault-k8s/1
   raft-cluster-bug-vault-k8s/2
   raft-cluster-bug-vault-k8s/4
   raft-cluster-bug-vault-k8s/3
Servers:
   raft-cluster-bug-vault-k8s/0
      Name:            raft-cluster-bug-vault-k8s/0
      Address:         10.1.166.210:8201
      Status:          leader
      Node Status:     alive
      Healthy:         true
      Last Contact:    0s
      Last Term:       3
      Last Index:      72
   raft-cluster-bug-vault-k8s/1
      Name:            raft-cluster-bug-vault-k8s/1
      Address:         10.1.166.193:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    494.230733ms
      Last Term:       3
      Last Index:      72
   raft-cluster-bug-vault-k8s/2
      Name:            raft-cluster-bug-vault-k8s/2
      Address:         10.1.166.224:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    99.573454ms
      Last Term:       3
      Last Index:      72
   raft-cluster-bug-vault-k8s/3
      Name:            raft-cluster-bug-vault-k8s/3
      Address:         10.1.166.255:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    4.714894438s
      Last Term:       3
      Last Index:      72
   raft-cluster-bug-vault-k8s/4
      Name:            raft-cluster-bug-vault-k8s/4
      Address:         10.1.166.220:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    2.488287328s
      Last Term:       3
      Last Index:      72
```

Delete a pod

```
kubectl delete pod vault-k8s-1 -n raft-cluster-bug
```

Wait for everything to settle back to active/idle. Observe everything still reports as healthy, even though the IP address is incorrect.

```console
$ vault operator raft autopilot state
Healthy:                      true
Failure Tolerance:            2
Leader:                       raft-cluster-bug-vault-k8s/0
Voters:
   raft-cluster-bug-vault-k8s/0
   raft-cluster-bug-vault-k8s/1
   raft-cluster-bug-vault-k8s/2
   raft-cluster-bug-vault-k8s/4
   raft-cluster-bug-vault-k8s/3
Servers:
   raft-cluster-bug-vault-k8s/0
      Name:            raft-cluster-bug-vault-k8s/0
      Address:         10.1.166.210:8201
      Status:          leader
      Node Status:     alive
      Healthy:         true
      Last Contact:    0s
      Last Term:       3
      Last Index:      74
   raft-cluster-bug-vault-k8s/1
      Name:            raft-cluster-bug-vault-k8s/1
      Address:         10.1.166.193:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    1.811061128s
      Last Term:       3
      Last Index:      73
   raft-cluster-bug-vault-k8s/2
      Name:            raft-cluster-bug-vault-k8s/2
      Address:         10.1.166.224:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    4.100232431s
      Last Term:       3
      Last Index:      74
   raft-cluster-bug-vault-k8s/3
      Name:            raft-cluster-bug-vault-k8s/3
      Address:         10.1.166.255:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    3.714476475s
      Last Term:       3
      Last Index:      74
   raft-cluster-bug-vault-k8s/4
      Name:            raft-cluster-bug-vault-k8s/4
      Address:         10.1.166.220:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    1.488745791s
      Last Term:       3
      Last Index:      74
```

Remove the leader

```bash
kubectl delete pod vault-k8s-0 -n raft-cluster-bug
```

Wait for things to settle. Now, things are reporting as unhealthy.

```console
$ vault operator raft autopilot state                                                                            2 ↵
Healthy:                      false
Failure Tolerance:            0
Leader:                       raft-cluster-bug-vault-k8s/2
Voters:
   raft-cluster-bug-vault-k8s/2
   raft-cluster-bug-vault-k8s/3
   raft-cluster-bug-vault-k8s/4
   raft-cluster-bug-vault-k8s/0
   raft-cluster-bug-vault-k8s/1
Servers:
   raft-cluster-bug-vault-k8s/0
      Name:            raft-cluster-bug-vault-k8s/0
      Address:         10.1.166.210:8201
      Status:          voter
      Node Status:     alive
      Healthy:         false
      Last Contact:    1m3.983504343s
      Last Term:       0
      Last Index:      0
   raft-cluster-bug-vault-k8s/1
      Name:            raft-cluster-bug-vault-k8s/1
      Address:         10.1.166.193:8201
      Status:          voter
      Node Status:     alive
      Healthy:         false
      Last Contact:    1m3.983503994s
      Last Term:       0
      Last Index:      0
   raft-cluster-bug-vault-k8s/2
      Name:            raft-cluster-bug-vault-k8s/2
      Address:         10.1.166.224:8201
      Status:          leader
      Node Status:     alive
      Healthy:         true
      Last Contact:    0s
      Last Term:       4
      Last Index:      101
   raft-cluster-bug-vault-k8s/3
      Name:            raft-cluster-bug-vault-k8s/3
      Address:         10.1.166.255:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    1.917555641s
      Last Term:       4
      Last Index:      101
   raft-cluster-bug-vault-k8s/4
      Name:            raft-cluster-bug-vault-k8s/4
      Address:         10.1.166.220:8201
      Status:          voter
      Node Status:     alive
      Healthy:         true
      Last Contact:    2.188396134s
      Last Term:       4
      Last Index:      101
```

#### Expected behavior
- The charms should display some error, if in fact the raft cluster is broken. 
  - But, raft itself seems to report everything as healthy at first. It's not obvious how we can detect the health of the raft cluster.

#### Environment

- Charm / library version (if relevant): 53 for reproducing, but the general issue applies beyond that.
- Juju version (output from `juju --version`): 3.1.6
- Cloud Environment: microk8s
- Kubernetes version (output from `kubectl version --short`):
```
Client Version: v1.27.7
Kustomize Version: v5.0.1
Server Version: v1.27.7
```

**Comment by saltiyazan** [comment]
As already mentioned in the ticket, it isn't possible to reproduce the issue with the same steps after the fix in #89 .
The charm after the fix uses hostnames instead of IP addresses so there will be no need to update IP addresses or Raft nodes. And based on a few tickets reported we see that there is a known issue with Raft nodes updating the IP addresses if the pod crashes and comes up again with a new IP.
Example issues:
- https://github.com/hashicorp/vault/issues/10702#issuecomment-769076069
- https://discuss.hashicorp.com/t/understanding-node-ip-changes-for-a-raft-cluster/56854
- https://discuss.hashicorp.com/t/change-vault-cluster-node-ip/14923/1
- https://github.com/hashicorp/vault/issues/17014

Based on fix #89 Raft cluster should not report unhealthy because of an IP change, however we still need to get a feedback if the cluster reports unhealthy for any other reason.

In general the cluster will report unhealthy if at least one of its nodes is reporting unhealthy and a node will report unhealthy it it can't heartbeat the leader unit or it can't keep up with the data replication from leader node (reference [here](https://developer.hashicorp.com/vault/docs/concepts/integrated-storage/autopilot#follower-health))

[PR](https://github.com/canonical/vault-k8s-operator/pull/117) #117 adds logging to report any node going unhealthy.

**Comment by saltiyazan** [comment]
PR #117 has been merged.
